tools/jsrun.py

```diff
@@ -33,7 +36,33 @@ def make_command(filename, engine=None, args=[]):
   # Use "'d8' in" because the name can vary, e.g. d8_g, d8, etc.
   return engine + [filename] + (['--expose-wasm', '--'] if 'd8' in jsengine or 'jsc' in jsengine else []) + args
 
-def run_js(filename, engine=None, args=[], check_timeout=False, stdin=None, stdout=PIPE, stderr=None, cwd=None, full_output=False, assert_returncode=0, error_limit=-1):
+
+def check_engine(engine):
+  if type(engine) is list:
+    engine_path = engine[0]
+  global WORKING_ENGINES
+  if engine_path in WORKING_ENGINES:
+    return WORKING_ENGINES[engine_path]
+  try:
+    logging.debug('Checking JS engine %s' % engine)
+    if 'hello, world!' in run_js(os.path.join(__rootpath__, 'src', 'hello_world.js'), engine, skip_check=True):
+      WORKING_ENGINES[engine_path] = True
+  except Exception, e:
+    logging.info('Checking JS engine %s failed. Check your config file. Details: %s' % (str(engine), str(e)))
+    WORKING_ENGINES[engine_path] = False
+  return WORKING_ENGINES[engine_path]
+
+
+def require_engine(engine):
+  engine_path = engine[0]
+  assert engine_path in WORKING_ENGINES, 'JS engine %s should have been checked at startup' % engine
```

Comment:
Yeah, the explicit checks of everything on startup made more sense with the cache. Without, it probably makes more sense just to be fully lazy.